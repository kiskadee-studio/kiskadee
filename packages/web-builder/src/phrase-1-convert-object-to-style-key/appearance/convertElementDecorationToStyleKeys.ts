import type {
  DecorationSchema,
  ClassNameMap,
  InteractionState,
  SolidColor,
  StyleKeyByElement
} from '@kiskadee/schema';
import { updateElementStyleKeyMap } from '../../utils';

/**
 * Gets a shadow property value for the given interaction state, falling back to 'rest' then to a
 * default.
 */
function getShadowValue<T>(
  property: Partial<Record<InteractionState, T>>,
  state: InteractionState,
  defaultValue: T
): T {
  return property[state] !== undefined
    ? property[state]
    : property.rest !== undefined
      ? property.rest
      : defaultValue;
}

/**
 * Converts an element's appearance settings into a map of style keys for each interaction state.
 *
 * Non-shadow properties (boolean, string, number, or array) generate a 'rest' state key:
 *   {property}__{value}
 * (arrays are JSON-stringified).
 *
 * Shadow properties (shadowX, shadowY, shadowBlur, shadowColor) are unified:
 *   - Determines all states present on any shadow property, plus 'rest'.
 *   - For each state, missing values fall back to 'rest', then to defaults
 *     (0 for offsets/blur, [0,0,0,1] for color).
 *   - Generates keys:
 *       rest:     shadow__[x,y,blur,[r,g,b,a]]
 *       non-rest: shadow--{state}__[x,y,blur,[r,g,b,a]]
 */
export function convertElementDecorationToStyleKeys(
  decoration: DecorationSchema
): StyleKeyByElement['decorations'] {
  const styleKeys: StyleKeyByElement['decorations'] = [];

  for (const [property, v] of Object.entries(decoration)) {
    if (property.startsWith('shadow') === false) {
      if (
        typeof v === 'boolean' ||
        typeof v === 'string' ||
        typeof v === 'number' ||
        Array.isArray(v)
      ) {
        const value = Array.isArray(v) ? JSON.stringify(v) : v;
        const styleKey = `${property}__${value}`;
        styleKeys.push(styleKey);
      }
    }
  }

  // CSS handles shadows as a single property despite having multiple attributes (x, y, blur, color).
  // So we need to combine all shadow properties into a unified value for each state.

  const hasShadowProperty =
    'shadowColor' in decoration ||
    'shadowX' in decoration ||
    'shadowY' in decoration ||
    'shadowBlur' in decoration;

  if (hasShadowProperty === true) {
    const { shadowX = {}, shadowY = {}, shadowBlur = {}, shadowColor = {} } = decoration;
    const allStates = new Set<InteractionState>([
      ...Object.keys(shadowX),
      ...Object.keys(shadowY),
      ...Object.keys(shadowBlur),
      ...Object.keys(shadowColor)
    ] as InteractionState[]);
    allStates.add('rest');

    for (const state of allStates) {
      const x = getShadowValue(shadowX, state, 0);
      const y = getShadowValue(shadowY, state, 0);
      const blur = getShadowValue(shadowBlur, state, 0);
      const color: SolidColor = getShadowValue(shadowColor, state, [0, 0, 0, 1]);
      const styleKey =
        state === 'rest'
          ? `shadow__[${x},${y},${blur},${JSON.stringify(color)}]`
          : `shadow--${state}__[${x},${y},${blur},${JSON.stringify(color)}]`;

      // elementStyleKeyMap = updateElementStyleKeyMap(
      //   elementStyleKeyMap,
      //   componentName,
      //   elementName,
      //   state,
      //   styleKey
      // );
      styleKeys.push(styleKey);
    }
  }

  return styleKeys;
}
