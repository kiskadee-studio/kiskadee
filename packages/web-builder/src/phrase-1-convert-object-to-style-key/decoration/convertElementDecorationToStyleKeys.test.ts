import type { DecorationSchema } from '@kiskadee/schema';
import { describe, expect, it } from 'vitest';
import { convertElementDecorationToStyleKeys } from './convertElementDecorationToStyleKeys';

describe('convertElementDecorationToStyleKeys', () => {
  describe('textFont style key generation', () => {
    it('should generate the textFont style key for a font family array', () => {
      const decoration: DecorationSchema = { textFont: ['Arial', 'Helvetica'] };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textFont__["Arial","Helvetica"]');
      expect(styleKey).toMatchSnapshot();
    });
  });

  describe('textItalic style key generation', () => {
    it('generates the textItalic__true style key when italic is enabled', () => {
      const decoration: DecorationSchema = { textItalic: true };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textItalic__true');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textItalic__false style key when italic is disabled', () => {
      const decoration: DecorationSchema = { textItalic: false };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textItalic__false');
      expect(styleKey).toMatchSnapshot();
    });
  });

  describe('textWeight style key generation', () => {
    it('generates the textWeight__thin style key for thin font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'thin' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__thin');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__extraLight style key for extraLight font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'extraLight' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__extraLight');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__light style key for light font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'light' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__light');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__normal style key for normal font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'normal' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__normal');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__medium style key for medium font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'medium' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__medium');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__semiBold style key for semiBold font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'semiBold' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__semiBold');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__bold style key for bold font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'bold' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__bold');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__extraBold style key for extraBold font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'extraBold' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__extraBold');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textWeight__black style key for black font weight', () => {
      const decoration: DecorationSchema = { textWeight: 'black' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textWeight__black');
      expect(styleKey).toMatchSnapshot();
    });
  });

  describe('textLineType style key generation', () => {
    it('generates the textLineType__none style key for none line type', () => {
      const decoration: DecorationSchema = { textLineType: 'none' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textLineType__none');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textLineType__underline style key for underline line type', () => {
      const decoration: DecorationSchema = { textLineType: 'underline' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textLineType__underline');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textLineType__lineThrough style key for lineThrough line type', () => {
      const decoration: DecorationSchema = { textLineType: 'lineThrough' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textLineType__lineThrough');
      expect(styleKey).toMatchSnapshot();
    });
  });

  describe('textAlign style key generation', () => {
    it('generates the textAlign__left style key for left text alignment', () => {
      const decoration: DecorationSchema = { textAlign: 'left' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textAlign__left');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textAlign__center style key for center text alignment', () => {
      const decoration: DecorationSchema = { textAlign: 'center' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textAlign__center');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the textAlign__right style key for right text alignment', () => {
      const decoration: DecorationSchema = { textAlign: 'right' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('textAlign__right');
      expect(styleKey).toMatchSnapshot();
    });
  });

  describe('borderStyle style key generation', () => {
    it('generates the borderStyle__none style key for none border style', () => {
      const decoration: DecorationSchema = { borderStyle: 'none' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('borderStyle__none');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the borderStyle__dotted style key for dotted border style', () => {
      const decoration: DecorationSchema = { borderStyle: 'dotted' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('borderStyle__dotted');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the borderStyle__dashed style key for dashed border style', () => {
      const decoration: DecorationSchema = { borderStyle: 'dashed' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('borderStyle__dashed');
      expect(styleKey).toMatchSnapshot();
    });

    it('generates the borderStyle__solid style key for solid border style', () => {
      const decoration: DecorationSchema = { borderStyle: 'solid' };
      const styleKey = convertElementDecorationToStyleKeys(decoration)[0];

      expect(styleKey).toEqual('borderStyle__solid');
      expect(styleKey).toMatchSnapshot();
    });
  });
});
