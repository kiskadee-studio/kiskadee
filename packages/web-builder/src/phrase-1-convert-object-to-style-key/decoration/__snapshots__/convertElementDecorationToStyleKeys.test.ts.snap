// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertElementDecorationToStyleKeys > borderStyle style key generation > generates the borderStyle__dashed style key for dashed border style 1`] = `"borderStyle__dashed"`;

exports[`convertElementDecorationToStyleKeys > borderStyle style key generation > generates the borderStyle__dotted style key for dotted border style 1`] = `"borderStyle__dotted"`;

exports[`convertElementDecorationToStyleKeys > borderStyle style key generation > generates the borderStyle__none style key for none border style 1`] = `"borderStyle__none"`;

exports[`convertElementDecorationToStyleKeys > borderStyle style key generation > generates the borderStyle__solid style key for solid border style 1`] = `"borderStyle__solid"`;

exports[`convertElementDecorationToStyleKeys > textAlign style key generation > generates the textAlign__center style key for center text alignment 1`] = `"textAlign__center"`;

exports[`convertElementDecorationToStyleKeys > textAlign style key generation > generates the textAlign__left style key for left text alignment 1`] = `"textAlign__left"`;

exports[`convertElementDecorationToStyleKeys > textAlign style key generation > generates the textAlign__right style key for right text alignment 1`] = `"textAlign__right"`;

exports[`convertElementDecorationToStyleKeys > textDecoration style key generation > generates the textDecoration__lineThrough style key for line-through text decoration 1`] = `"textDecoration__lineThrough"`;

exports[`convertElementDecorationToStyleKeys > textDecoration style key generation > generates the textDecoration__none style key for none text decoration 1`] = `"textDecoration__none"`;

exports[`convertElementDecorationToStyleKeys > textDecoration style key generation > generates the textDecoration__underline style key for underline text decoration 1`] = `"textDecoration__underline"`;

exports[`convertElementDecorationToStyleKeys > textFont style key generation > should generate the textFont style key for a font family array 1`] = `"textFont__["Arial","Helvetica"]"`;

exports[`convertElementDecorationToStyleKeys > textFontFamily style key generation > generates the textItalic__true style key when italic is enabled 1`] = `"textFontFamily__["Arial","Helvetica"]"`;

exports[`convertElementDecorationToStyleKeys > textFontFamily style key generation > should generate the textFontFamily style key for a font family array 1`] = `"textFontFamily__["Arial","Helvetica"]"`;

exports[`convertElementDecorationToStyleKeys > textItalic style key generation > generates the textItalic__false style key when italic is disabled 1`] = `"textItalic__false"`;

exports[`convertElementDecorationToStyleKeys > textItalic style key generation > generates the textItalic__true style key when italic is enabled 1`] = `"textItalic__true"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__black style key for black font weight 1`] = `"textWeight__black"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__bold style key for bold font weight 1`] = `"textWeight__bold"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__extraBold style key for extraBold font weight 1`] = `"textWeight__extraBold"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__extraLight style key for extraLight font weight 1`] = `"textWeight__extraLight"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__light style key for light font weight 1`] = `"textWeight__light"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__medium style key for medium font weight 1`] = `"textWeight__medium"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__normal style key for normal font weight 1`] = `"textWeight__normal"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__semiBold style key for semiBold font weight 1`] = `"textWeight__semiBold"`;

exports[`convertElementDecorationToStyleKeys > textWeight style key generation > generates the textWeight__thin style key for thin font weight 1`] = `"textWeight__thin"`;
